packages/models/src/store/interface.ts

```diff
@@ -71,31 +71,60 @@ type PickExist<T, K extends keyof T> = OmitByValue<{
 
 type GetScriptOption<T extends ScriptSchema> = IsEmptyScriptSchemaNever<
   OmitByValue<{
-    [P in keyof T]: T[P] extends { offset?: number; length?: number; schema: unknown }
+    [P in keyof T]: T[P] extends { offset?: ByteLength; length?: ByteLength; schema: unknown }
       ? PickExist<T[P], 'offset' | 'length'>
-      : never
+      : true
   }>
 >
 
 export type GetStorageOption<T extends StorageSchema> = IfEmptyStorageSchemaNever<
   OmitByValue<{
-    data: T extends { data: infer Option extends { offset?: number; length?: number; schema: unknown } }
-      ? PickExist<Option, 'offset' | 'length'>
+    [P in keyof T]: P extends 'data'
+      ? T extends { data: infer Option extends { offset?: ByteLength; length?: ByteLength; schema: unknown } }
+        ? PickExist<Option, 'offset' | 'length'>
+        : true
+      : P extends 'witness'
+      ? T extends { witness: infer Option extends { offset?: ByteLength; length?: ByteLength; schema: unknown } }
+        ? PickExist<Option, 'offset' | 'length'>
+        : true
+      : P extends 'lock'
+      ? T extends { lock: infer Lock extends ScriptSchema }
+        ? GetScriptOption<Lock>
+        : never
+      : P extends 'type'
+      ? T extends { type: infer Type extends ScriptSchema }
+        ? GetScriptOption<Type>
+        : never
       : never
-    witness: T extends { witness: infer Option extends { offset?: number; length?: number; schema: unknown } }
-      ? PickExist<Option, 'offset' | 'length'>
-      : never
-    lock: T extends { lock: infer Lock extends ScriptSchema } ? GetScriptOption<Lock> : never
-    type: T extends { type: infer Type extends ScriptSchema } ? GetScriptOption<Type> : never
   }>
 >
 
+export type GetOnChainStorage<T extends StorageSchema> = {
+  data: IsKeysExist<T, 'data'> extends true ? string : never
+  witness: IsKeysExist<T, 'witness'> extends true ? string : never
+  lock: T extends { lock: infer Lock extends ScriptSchema }
+    ? IsEmptyScriptSchemaNever<{ [P in keyof Lock]: string }>
+    : never
+  type: T extends { type: infer Type extends ScriptSchema }
+    ? IsEmptyScriptSchemaNever<{ [P in keyof Type]: string }>
+    : never
+}
+
 export type StorageLocation = 'data' | 'witness' | ['lock' | 'type', keyof ScriptSchema]
 
 export type StorePath<K = StorageLocation> = K extends string[] ? [...K, ...string[]] : [K, ...string[]]
 
-export interface StoreMessage<State> {
-  type: 'add_state' | 'remove_state'
-  add?: Record<OutPointString, State>
-  remove?: OutPointString[]
+export type UpdateStorageValue = {
+  witness: HexString
+  cell: Cell
 }
+
+export type StoreMessage =
+  | {
+      type: 'remove_cell'
+      value: OutPointString[]
+    }
+  | {
+      type: 'update_cell'
+      value: UpdateStorageValue | UpdateStorageValue[]
```

Comment:
I would suggest defining 
```
{ type: 'update_cells`, value: UpdateStorageValue[] }
```
Because it covers `UpdateStorageValue | UpdateStorageValue[]` and `update_cells` is clearly for multiple cells.